File: lib/data/repositories/species_repository.dart
import 'package:drift/drift.dart';

import '../../domain/models/species.dart';
import '../../infra/db/app_database.dart';
import '../../infra/db/tables.dart';

class SpeciesRepository {
  final AppDatabase _db;
  SpeciesRepository(this._db);

  Future<int> insert({required String name, String? iconKey}) async {
    return _db.into(_db.speciesTable).insert(
          SpeciesTableCompanion.insert(
            name: name,
            iconKey: Value(iconKey),
            createdAt: DateTime.now().millisecondsSinceEpoch,
          ),
        );
  }

  Future<void> rename({required int id, required String name}) async {
    await (_db.update(_db.speciesTable)..where((s) => s.id.equals(id)))
        .write(SpeciesTableCompanion(name: Value(name)));
  }

  Future<void> archive(int id) async {
    await (_db.update(_db.speciesTable)..where((s) => s.id.equals(id))).write(
      SpeciesTableCompanion(
        archivedAt: Value(DateTime.now().millisecondsSinceEpoch),
      ),
    );
  }

  Future<void> unarchive(int id) async {
    await (_db.update(_db.speciesTable)..where((s) => s.id.equals(id)))
        .write(const SpeciesTableCompanion(archivedAt: Value(null)));
  }

  Future<List<Species>> listActive() async {
    final rows = await (_db.select(_db.speciesTable)
          ..where((s) => s.archivedAt.isNull())
          ..orderBy([(s) => OrderingTerm(expression: s.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<List<Species>> listArchived() async {
    final rows = await (_db.select(_db.speciesTable)
          ..where((s) => s.archivedAt.isNotNull())
          ..orderBy([(s) => OrderingTerm(expression: s.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<List<Species>> listAll() async {
    final rows = await (_db.select(_db.speciesTable)
          ..orderBy([(s) => OrderingTerm(expression: s.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<int> countActive() async {
    final rows = await (_db.select(_db.speciesTable)
          ..where((s) => s.archivedAt.isNull()))
        .get();
    return rows.length;
  }

  Species _toDomain(SpeciesRow row) => Species(
        id: row.id,
        name: row.name,
        iconKey: row.iconKey,
        archivedAt: row.archivedAt == null
            ? null
            : DateTime.fromMillisecondsSinceEpoch(row.archivedAt!),
      );
}

File: src/data/repositories/species-repository.ts
import { asc, eq, isNull, isNotNull } from 'drizzle-orm';
import { species } from '@/infra/db/schema';
import type { Db } from '@/infra/db/client';
import { Species } from '@/domain/models/species';

interface SpeciesRow {
  id: string;
  label: string;
  iconKey: string | null;
  ordering: number;
  isCustom: number;
  archivedAt: string | null;
}

function toRow(s: Species) {
  return { ...s, isCustom: s.isCustom ? 1 : 0 };
}
function fromRow(r: SpeciesRow): Species {
  return Species.parse({ ...r, isCustom: r.isCustom === 1 });
}

export class SpeciesRepository {
  constructor(private readonly db: Db) {}

  async byId(id: string): Promise<Species | null> {
    const rows = await this.db.select().from(species).where(eq(species.id, id));
    return rows[0] ? fromRow(rows[0] as SpeciesRow) : null;
  }
  async listAll(): Promise<Species[]> {
    const rows = await this.db.select().from(species).orderBy(asc(species.ordering));
    return rows.map((r) => fromRow(r as SpeciesRow));
  }
  async listActive(): Promise<Species[]> {
    const rows = await this.db
      .select()
      .from(species)
      .where(isNull(species.archivedAt))
      .orderBy(asc(species.ordering));
    return rows.map((r) => fromRow(r as SpeciesRow));
  }
  async listArchived(): Promise<Species[]> {
    const rows = await this.db
      .select()
      .from(species)
      .where(isNotNull(species.archivedAt))
      .orderBy(asc(species.ordering));
    return rows.map((r) => fromRow(r as SpeciesRow));
  }
  async upsert(s: Species): Promise<void> {
    const row = toRow(s);
    await this.db.insert(species).values(row).onConflictDoUpdate({ target: species.id, set: row });
  }
  async delete(id: string): Promise<void> {
    await this.db.delete(species).where(eq(species.id, id));
  }
}
